src/Microsoft.AspNet.Authentication.OAuth/OAuthHandler.cs

```diff
@@ -30,131 +29,71 @@ public OAuthHandler(HttpClient backchannel)
 
         protected HttpClient Backchannel { get; private set; }
 
-        public override async Task<bool> InvokeAsync()
+        protected override async Task<AuthenticateResult> HandleRemoteAuthenticateAsync()
         {
-            if (Options.CallbackPath.HasValue && Options.CallbackPath == Request.Path)
-            {
-                return await InvokeReturnPathAsync();
-            }
-            return false;
-        }
+            AuthenticationProperties properties = null;
+            var query = Request.Query;
 
-        public async Task<bool> InvokeReturnPathAsync()
-        {
-            var ticket = await HandleAuthenticateOnceAsync();
-            if (ticket == null)
+            var error = query["error"];
+            if (!StringValues.IsNullOrEmpty(error))
             {
-                Logger.LogWarning("Invalid return state, unable to redirect.");
-                Response.StatusCode = 500;
-                return true;
+                return AuthenticateResult.Failed(error);
             }
 
-            var context = new SigningInContext(Context, ticket)
-            {
-                SignInScheme = Options.SignInScheme,
-                RedirectUri = ticket.Properties.RedirectUri,
-            };
-            ticket.Properties.RedirectUri = null;
-
-            await Options.Events.SigningIn(context);
+            var code = query["code"];
+            var state = query["state"];
 
-            if (context.SignInScheme != null && context.Principal != null)
+            properties = Options.StateDataFormat.Unprotect(state);
+            if (properties == null)
             {
-                await Context.Authentication.SignInAsync(context.SignInScheme, context.Principal, context.Properties);
+                return AuthenticateResult.Failed("State not found.");
```

Comment:
Or unprotect failed (and returned null). "The oauth state was missing or invalid."